[2.x] Inertia.js: Add support for all Jetstream-provided views.

claudiodekker opened the pull request:
This PR adds Inertia variants for all Jetstream included views. This allows a more native Inertia experience for users of this stack, and solves several issues and difficulties that makes for an all-in-all smoother experience.

For example, some Jetstream users have been asking questions on our Discord regarding difficulties they're experiencing with getting Jetstream to play well with Inertia. Specifically, the most common issue is when a session expires, as the user then gets redirected to the login page, and because that login page is a Blade view, Inertia renders it in a so-called 'error modal'.

To solve or prevent this, users should instead use Inertia for their login pages as well, and while some have gotten to the point where they're attempting to (re)configure Fortify to use Inertia views instead, this too isn't the most straight-forward thing to do.

reinink commented:
@taylorotwell For what it's worth, I definitely think this is the right approach to go. The best Inertia.js experience is to do your whole app with Vue based page components. Session expiration "just works" this way, since users are automatically redirected back to the login page, which itself is an Inertia endpoint. I don't think I'd ever, personally, build an app that had a mix of Blade views and Vue page components. That said, this does create an extra maintenance burden on this project, since these auth views can no longer be shared between the two stacks. I feel like that's a worthwhile tradeoff though. 🤷‍♂️ 

That all said, the session expiration issue could be solved using an `Inertia::location($url)` (409) response that's called in the authentication middleware. This would force a "full page redirect" to the login page. To me this is more of a band-aid, and I'd again prefer the solution in this PR. 👍

@claudiodekker Excellent work brother, this looks great. 🙌

claudiodekker commented:
Thanks! 

On a side-note (just so that I don't forget), Once https://github.com/laravel/fortify/pull/107 gets tagged, I can remove the `->toResponse($request)` calls from this PR as well. Since it hasn't yet, I've decided to leave them in for now.

reinink commented:
Yup, and while we're creating todo lists, it would be great to get `@inertiajs/inertia` on `v0.3.0`, and then update all the Jetstream code to use the new event callbacks. When we make that change, we should add `@inertiajs/progress` to the project, since right now there isn't any loading indicators, since the NProgress CSS isn't included.

taylorotwell commented:
@claudiodekker I've just tagged a new laravel/fortify release.

claudiodekker commented:
@taylorotwell Thanks! I've updated the PR. I also updated the `composer.json` to reflect the minimum Fortify version to be `^1.6.1`, let me know if you'd rather leave that unchanged.

m1guelpf commented on stubs/inertia/resources/js/Jetstream/AuthenticationCardLogo.vue:
```diff
@@ -0,0 +1,8 @@
+<template>
+    <a href="/">
+        <svg class="w-16 h-16" viewBox="0 0 48 48" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Couldn't we just reuse the `ApplicationMark` component here instead of duplicating?

claudiodekker replied:
Yeah, I looked at that, but it's different, and because I'm trying to replicate the other views to keep branding consistent I copied this one instead.

taylorotwell commented:
@claudiodekker  Does this need to be rebased or something? It seems to contain some random other commits?

claudiodekker commented:
@taylorotwell Yeah, I merged 1.x into it.. I think that's not yet merged back in to master.

taylorotwell commented:
I've merged 1.x into master.

claudiodekker commented:
@taylorotwell Fixed, thanks! 👍

taylorotwell commented:
Sick. Will probably review and get this merged in tomorrow.

claudiodekker commented:
Cool, let me know if there's anything else you want me to do 👍

taylorotwell commented:
@claudiodekker One bug I've noticed so far... on two-factor code entry screen after login... entering the code and then hitting "enter" just toggles you to recovery code form instead of submitting auth code.

claudiodekker commented:
Oof, yeah, that shouldn't happen. 

I'll look at it in a bit, and will also double check the keyboard behavior of other forms.

taylorotwell commented:
That was the only thing I hit so far. I think I've tested every form.

claudiodekker commented:
Great, thanks for letting me know! 🙏 

Having dinner right now, so it might take an hour or so before I have a fix. Want me to ping you when it's done?

taylorotwell commented:
I think I'l have it fixed. Don't worry about it.

m1guelpf commented on stubs/inertia/resources/js/Pages/Auth/TwoFactorChallenge.vue:
```diff
@@ -0,0 +1,89 @@
+        },
+        methods: {
+            toggleRecovery() {
+                this.recovery ^= true
```
```suggestion
                this.recovery = true
```

taylorotwell replied:
They are using this to toggle between 0 and 1.

m1guelpf replied:
Wouldn't it be less confusing to use a boolean?

claudiodekker replied:
This is actually correct. It's a bitwise XOR that changes true to false, and false to true in this situation (hence toggle) 👍

m1guelpf replied:
That's actually a neat trick!

claudiodekker replied:
For sure :) Also works in PHP ;)

m1guelpf commented on stubs/inertia/resources/js/Pages/Auth/TwoFactorChallenge.vue:
```diff
@@ -0,0 +1,89 @@
+            </div>
+
+            <div class="flex items-center justify-end mt-4">
+                <button @click.prevent="toggleRecovery" class="text-sm text-gray-600 hover:text-gray-900 underline cursor-pointer">
```
```suggestion
                <button type="button" @click.prevent="toggleRecovery" class="text-sm text-gray-600 hover:text-gray-900 underline cursor-pointer">
```

m1guelpf commented on stubs/inertia/resources/js/Pages/Auth/TwoFactorChallenge.vue:
```diff
@@ -0,0 +1,89 @@
+                        Use an authentication code
+                    </template>
+                </button>
+                <jet-button class="ml-4" :class="{ 'opacity-25': form.processing }" :disabled="form.processing">
```
```suggestion
                <jet-button type="submit" class="ml-4" :class="{ 'opacity-25': form.processing }" :disabled="form.processing">
```

taylorotwell replied:
Already the default type for jet buttons.
